File: InputController.m
/*
Copyright (C) 2010-2012 Kristian Duske

This file is part of TrenchBroom.

TrenchBroom is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

TrenchBroom is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with TrenchBroom.  If not, see <http://www.gnu.org/licenses/>.
*/

#import "InputController.h"
#import <OpenGL/glu.h>
#import "CameraTool.h"
#import "SelectionTool.h"
#import "MoveTool.h"
#import "RotateTool.h"
#import "FaceTool.h"
#import "ClipTool.h"
#import "Camera.h"
#import "Picker.h"
#import "PickingHit.h"
#import "PickingHitList.h"
#import "MapView3D.h"
#import "SelectionManager.h"
#import "Face.h"
#import "Brush.h"
#import "MapWindowController.h"
#import "MapDocument.h"
#import "MoveTool.h"
#import "FaceTool.h"
#import "DragVertexTool.h"
#import "DragEdgeTool.h"
#import "DragFaceTool.h"
#import "Options.h"
#import "DefaultFilter.h"
#import "DndTool.h"
#import "EntityDefinitionDndTool.h"
#import "EntityDefinitionManager.h"
#import "EntityDefinition.h"
#import "Entity.h"
#import "CreateBrushTool.h"

@interface InputController (private)

- (void)updateEvent:(NSEvent *)event;
- (void)updateRay;
- (void)cameraViewChanged:(NSNotification *)notification;
- (void)mapChanged:(NSNotification *)notification;
- (void)selectionChanged:(NSNotification *)notification;

- (void)showContextMenu;
@end

@implementation InputController (private)

- (void)updateEvent:(NSEvent *)event {
    [lastEvent release];
    lastEvent = [event retain];
}

- (void)updateRay {
    MapView3D* mapView3D = [windowController view3D];
    Camera* camera = [windowController camera];
    [[mapView3D openGLContext] makeCurrentContext];
    
    NSPoint m = [mapView3D convertPointFromBase:[lastEvent locationInWindow]];
    lastRay = [camera pickRayX:m.x y:m.y];
    
    [currentHits release];
    currentHits = nil;
}

- (void)cameraViewChanged:(NSNotification *)notification {
    [self updateRay];
}

- (void)mapChanged:(NSNotification *)notification {
    [currentHits release];
    currentHits = nil;
}

- (void)showContextMenu {
    if (popupMenu == nil) {
        popupMenu = [[windowController view3D] menu];
        
        NSMenu* pointEntityMenu = [[windowController view3D] pointEntityMenu];
        NSMenu* brushEntityMenu = [[windowController view3D] brushEntityMenu];
        
        EntityDefinitionManager* entityDefinitionManager = [[windowController document] entityDefinitionManager];
        
        NSArray* pointDefinitions = [entityDefinitionManager definitionsOfType:EDT_POINT];
        for (int i = 0; i < [pointDefinitions count]; i++) {
            EntityDefinition* definition = [pointDefinitions objectAtIndex:i];
            NSMenuItem* definitionItem = [pointEntityMenu insertItemWithTitle:[definition name] action:@selector(createPointEntity:) keyEquivalent:@"" atIndex:i];
            [definitionItem setTag:i];
        }
        
        NSArray* brushDefinitions = [entityDefinitionManager definitionsOfType:EDT_BRUSH];
        for (int i = 0; i < [brushDefinitions count]; i++) {
            EntityDefinition* definition = [brushDefinitions objectAtIndex:i];
            NSMenuItem* definitionItem = [brushEntityMenu insertItemWithTitle:[definition name] action:@selector(createBrushEntity:) keyEquivalent:@"" atIndex:i];
            [definitionItem setTag:i];
        }
    }
    
    menuPosition = [lastEvent locationInWindow];
    menuPosition = [[windowController view3D] convertPointFromBase:menuPosition];
    [NSMenu popUpContextMenu:popupMenu withEvent:lastEvent forView:[windowController view3D]];
}

- (void)selectionChanged:(NSNotification *)notification {
    SelectionManager* selectionManager = [windowController selectionManager];
    if (![selectionManager hasSelectedBrushes] && modalReceiverIndex != -1) {
        Tool* modalReceiver = [receiverChain objectAtIndex:modalReceiverIndex];
        [modalReceiver deactivated:lastEvent ray:&lastRay hits:[self currentHits]];
        [receiverChain removeObjectAtIndex:modalReceiverIndex];
        modalReceiverIndex = -1;
    }
}

@end

@implementation InputController
- (id)initWithWindowController:(MapWindowController *)theWindowController {
    if ((self = [self init])) {
        windowController = theWindowController; // do not retain
        MapDocument* map = [windowController document];
        
        SelectionManager* selectionManager = [map selectionManager];
        GroupManager* groupManager = [map groupManager];
        Options* options = [windowController options];
        Camera* camera = [windowController camera];
        
        filter = [[DefaultFilter alloc] initWithSelectionManager:selectionManager groupManager:groupManager camera:camera options:options];
        
        cameraTool = [[CameraTool alloc] initWithWindowController:windowController];
        selectionTool = [[SelectionTool alloc] initWithWindowController:windowController];
        moveTool = [[MoveTool alloc] initWithWindowController:windowController];
        createBrushTool = [[CreateBrushTool alloc] initWithWindowController:windowController];
        rotateTool = [[RotateTool alloc] initWithWindowController:windowController];
        faceTool = [[FaceTool alloc] initWithWindowController:windowController];
        dragVertexTool = [[DragVertexTool alloc] initWithWindowController:windowController];
        dragEdgeTool = [[DragEdgeTool alloc] initWithWindowController:windowController];
        dragFaceTool = [[DragFaceTool alloc] initWithWindowController:windowController];
        clipTool = [[ClipTool alloc] initWithWindowController:windowController];
        entityDefinitionDndTool = [[EntityDefinitionDndTool alloc] initWithWindowController:windowController];
        
        receiverChain = [[NSMutableArray alloc] init];
        [receiverChain addObject:cameraTool];
        [receiverChain addObject:faceTool];
        [receiverChain addObject:moveTool];
        [receiverChain addObject:rotateTool];
        [receiverChain addObject:selectionTool];
        [receiverChain addObject:createBrushTool];
        dragScrollReceiver = nil;
        modalReceiverIndex = -1;
        
        NSNotificationCenter* center = [NSNotificationCenter defaultCenter];
        [center addObserver:self selector:@selector(cameraViewChanged:) name:CameraViewChanged object:camera];
        [center addObserver:self selector:@selector(mapChanged:) name:EntitiesAdded object:map];
        [center addObserver:self selector:@selector(mapChanged:) name:EntitiesWereRemoved object:map];
        [center addObserver:self selector:@selector(mapChanged:) name:PropertiesDidChange object:map];
        [center addObserver:self selector:@selector(mapChanged:) name:BrushesAdded object:map];
        [center addObserver:self selector:@selector(mapChanged:) name:BrushesWereRemoved object:map];
        [center addObserver:self selector:@selector(mapChanged:) name:BrushesDidChange object:map];
        [center addObserver:self selector:@selector(mapChanged:) name:FacesDidChange object:map];
        [center addObserver:self selector:@selector(selectionChanged:) name:SelectionRemoved object:selectionManager];
    }
    
    return self;
}


- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    [entityDefinitionDndTool release];
    [cameraTool release];
    [selectionTool release];
    [moveTool release];
    [createBrushTool release];
    [rotateTool release];
    [faceTool release];
    [dragVertexTool release];
    [dragEdgeTool release];
    [dragFaceTool release];
    [clipTool release];
    [lastEvent release];
    [currentHits release];
    [receiverChain release];
    [filter release];
    [super dealloc];
}

- (BOOL)keyDown:(NSEvent *)event sender:(id)sender {
    switch ([event keyCode]) {
        case 48: // tab key, faking this because it doesn't work as a key equivalent for NSMenuItem
            if (clipTool != nil) {
                [clipTool toggleClipSide];
                return YES;
            }
            break;
        default:
            break;
    }
    
    return NO;
}

- (BOOL)keyUp:(NSEvent *)event sender:(id)sender {
    return NO;
}

- (void)flagsChanged:(NSEvent *)event sender:(id)sender {
    for (Tool* receiver in receiverChain)
        [receiver flagsChanged:event ray:&lastRay hits:[self currentHits]];
}

- (void)leftMouseDragged:(NSEvent *)event sender:(id)sender {
    if (dragStatus == MS_NONE) {
        dragStatus = MS_LEFT;
        
        for (Tool* receiver in receiverChain) {
            if ([receiver beginLeftDrag:lastEvent ray:&lastRay hits:[self currentHits]]) {
                dragScrollReceiver = receiver;
                break;
            }
        }
    }

    [self updateEvent:event];
    [self updateRay];
    [dragScrollReceiver leftDrag:lastEvent ray:&lastRay hits:[self currentHits]];
}

- (void)mouseMoved:(NSEvent *)event sender:(id)sender {
    [self updateEvent:event];
    [self updateRay];

    for (Tool* receiver in receiverChain)
        [receiver mouseMoved:lastEvent ray:&lastRay hits:[self currentHits]];
}

- (void)mouseEntered:(NSEvent *)event sender:(id)sender {
    hasMouse = YES;
    [self updateEvent:event];
    [self updateRay];
}

- (void)mouseExited:(NSEvent *)event sender:(id)sender {
    hasMouse = NO;
}

- (void)leftMouseDown:(NSEvent *)event sender:(id)sender {
    [self updateEvent:event];
    [self updateRay];
    
    for (Tool* receiver in receiverChain) {
        if ([receiver leftMouseDown:lastEvent ray:&lastRay hits:[self currentHits]])
            break;
    }
}

- (void)leftMouseUp:(NSEvent *)event sender:(id)sender {
    [self updateEvent:event];
    
    if (dragStatus == MS_LEFT || scrollStatus == MS_LEFT) {
        if (dragStatus == MS_LEFT) {
            [dragScrollReceiver endLeftDrag:lastEvent ray:&lastRay hits:[self currentHits]];
            dragStatus = MS_NONE;
            dragScrollReceiver = nil;
        }
        
        if (scrollStatus == MS_LEFT) {
            [dragScrollReceiver endLeftScroll:lastEvent ray:&lastRay hits:[self currentHits]];
            scrollStatus = MS_NONE;
            dragScrollReceiver = nil;
        }
        NSAssert([[[windowController document] undoManager] groupingLevel] == 0, @"undo grouping level must be 0 after drag ended");
    } else {
        for (Tool* receiver in receiverChain) {
            if ([receiver leftMouseUp:lastEvent ray:&lastRay hits:[self currentHits]])
                break;
        }
    }
}

- (void)rightMouseDragged:(NSEvent *)event sender:(id)sender {
    if (dragStatus == MS_NONE) {
        dragStatus = MS_RIGHT;
        for (Tool* receiver in receiverChain) {
            if ([receiver beginRightDrag:lastEvent ray:&lastRay hits:[self currentHits]]) {
                dragScrollReceiver = receiver;
                break;
            }
        }
    }
    
    [self updateEvent:event];
    [self updateRay];
    [dragScrollReceiver rightDrag:lastEvent ray:&lastRay hits:[self currentHits]];
}

- (void)rightMouseDown:(NSEvent *)event sender:(id)sender {
    [self updateEvent:event];
    
    for (Tool* receiver in receiverChain) {
        if ([receiver rightMouseDown:lastEvent ray:&lastRay hits:[self currentHits]])
            break;
    }
}

- (void)rightMouseUp:(NSEvent *)event sender:(id)sender {
    [self updateEvent:event];
    
    if (dragStatus == MS_RIGHT || scrollStatus == MS_RIGHT) {
        if (dragStatus == MS_RIGHT) {
            [dragScrollReceiver endRightDrag:lastEvent ray:&lastRay hits:[self currentHits]];
            dragStatus = MS_NONE;
            dragScrollReceiver = nil;
        }
        
        if (scrollStatus == MS_RIGHT) {
            [dragScrollReceiver endRightScroll:lastEvent ray:&lastRay hits:[self currentHits]];
            scrollStatus = MS_NONE;
            dragScrollReceiver = nil;
        }
    } else {
        for (Tool* receiver in receiverChain) {
            if ([receiver rightMouseUp:lastEvent ray:&lastRay hits:[self currentHits]])
                break;
        }
        if (dragStatus == MS_NONE)
            [self showContextMenu];
    }
}

- (void)scrollWheel:(NSEvent *)event sender:(id)sender {
    int buttons = [NSEvent pressedMouseButtons];
    if (scrollStatus == MS_NONE && (buttons == 1 || buttons == 2)) {
        scrollStatus = buttons == 1 ? MS_LEFT : MS_RIGHT;
        if (scrollStatus == MS_LEFT) {
            if (dragScrollReceiver == nil) {
                for (Tool* receiver in receiverChain) {
                    if ([receiver beginLeftScroll:lastEvent ray:&lastRay hits:[self currentHits]]) {
                        dragScrollReceiver = receiver;
                        break;
                    }
                }
            } else {
                [dragScrollReceiver beginLeftScroll:lastEvent ray:&lastRay hits:[self currentHits]];
            }
        } else {
            if (dragScrollReceiver == nil) {
                for (Tool* receiver in receiverChain) {
                    if ([receiver beginRightScroll:lastEvent ray:&lastRay hits:[self currentHits]]) {
                        dragScrollReceiver = receiver;
                        break;
                    }
                }
            } else {
                [dragScrollReceiver beginRightScroll:lastEvent ray:&lastRay hits:[self currentHits]];
            }
        }
    }

    [self updateEvent:event];
    if (dragScrollReceiver != nil) {
        if (scrollStatus == MS_LEFT)
            [dragScrollReceiver leftScroll:lastEvent ray:&lastRay hits:[self currentHits]];
        else if (scrollStatus == MS_RIGHT)
            [dragScrollReceiver rightScroll:lastEvent ray:&lastRay hits:[self currentHits]];
        else
            [dragScrollReceiver scrollWheel:lastEvent ray:&lastRay hits:[self currentHits]];
    } else {
        for (Tool* receiver in receiverChain) {
            if ([receiver scrollWheel:lastEvent ray:&lastRay hits:[self currentHits]])
                break;
        }
    }
}

- (void)beginGesture:(NSEvent *)event sender:(id)sender {
    [self updateEvent:event];

    for (Tool* receiver in receiverChain) {
        if ([receiver beginGesture:lastEvent ray:&lastRay hits:[self currentHits]]) {
            dragScrollReceiver = receiver;
            break;
        }
    }
}

- (void)endGesture:(NSEvent *)event sender:(id)sender {
    NSAssert(dragScrollReceiver != nil, @"modal receiver must not be nil");
    [self updateEvent:event];
    [dragScrollReceiver endGesture:lastEvent ray:&lastRay hits:[self currentHits]];
}

- (void)magnify:(NSEvent *)event sender:(id)sender {
    NSAssert(dragScrollReceiver != nil, @"modal receiver must not be nil");
    [self updateEvent:event];
    [dragScrollReceiver magnify:lastEvent ray:&lastRay hits:[self currentHits]];
}

- (NSDragOperation)draggingEntered:(id <NSDraggingInfo>)sender {
    NSPasteboard* pasteboard = [sender draggingPasteboard];
    NSString* type = [pasteboard availableTypeFromArray:[NSArray arrayWithObject:EntityDefinitionType]];
    if (type == nil)
        return NSDragOperationNone;

    if ([EntityDefinitionType isEqualToString:type])
        activeDndTool = entityDefinitionDndTool;
    
    if (activeDndTool != nil) {
        Camera* camera = [windowController camera];
        NSPoint location = [sender draggingLocation];
        TRay ray = [camera pickRayX:location.x y:location.y];
        
        Picker* picker = [[windowController document] picker];
        PickingHitList* hitList = [picker pickObjects:&ray filter:filter];
        
        return [activeDndTool handleDraggingEntered:sender ray:&ray hits:hitList];
    }
    
    return NSDragOperationNone;
}

- (NSDragOperation)draggingUpdated:(id <NSDraggingInfo>)sender {
    if (activeDndTool == nil)
        return NSDragOperationNone;
    
    Camera* camera = [windowController camera];
    NSPoint location = [sender draggingLocation];
    TRay ray = [camera pickRayX:location.x y:location.y];
    
    Picker* picker = [[windowController document] picker];
    PickingHitList* hitList = [picker pickObjects:&ray filter:filter];
    
    return [activeDndTool handleDraggingUpdated:sender ray:&ray hits:hitList];
}

- (void)draggingEnded:(id <NSDraggingInfo>)sender {
    if (activeDndTool == nil)
        return;
    
    Camera* camera = [windowController camera];
    NSPoint location = [sender draggingLocation];
    TRay ray = [camera pickRayX:location.x y:location.y];
    
    Picker* picker = [[windowController document] picker];
    PickingHitList* hitList = [picker pickObjects:&ray filter:filter];
    
    [activeDndTool handleDraggingEnded:sender ray:&ray hits:hitList];
    activeDndTool = nil;
}

- (void)draggingExited:(id <NSDraggingInfo>)sender {
    if (activeDndTool == nil)
        return;
    
    Camera* camera = [windowController camera];
    NSPoint location = [sender draggingLocation];
    TRay ray = [camera pickRayX:location.x y:location.y];
    
    Picker* picker = [[windowController document] picker];
    PickingHitList* hitList = [picker pickObjects:&ray filter:filter];
    
    [activeDndTool handleDraggingExited:sender ray:&ray hits:hitList];
    activeDndTool = nil;
}

- (BOOL)prepareForDragOperation:(id <NSDraggingInfo>)sender {
    if (activeDndTool == nil)
        return NO;
    
    Camera* camera = [windowController camera];
    NSPoint location = [sender draggingLocation];
    TRay ray = [camera pickRayX:location.x y:location.y];
    
    Picker* picker = [[windowController document] picker];
    PickingHitList* hitList = [picker pickObjects:&ray filter:filter];
    
    return [activeDndTool prepareForDragOperation:sender ray:&ray hits:hitList];
}

- (BOOL)performDragOperation:(id <NSDraggingInfo>)sender {
    if (activeDndTool == nil)
        return NO;
    
    Camera* camera = [windowController camera];
    NSPoint location = [sender draggingLocation];
    TRay ray = [camera pickRayX:location.x y:location.y];
    
    Picker* picker = [[windowController document] picker];
    PickingHitList* hitList = [picker pickObjects:&ray filter:filter];
    
    return [activeDndTool performDragOperation:sender ray:&ray hits:hitList];
}

- (void)concludeDragOperation:(id <NSDraggingInfo>)sender {
    if (activeDndTool == nil)
        return;
    
    Camera* camera = [windowController camera];
    NSPoint location = [sender draggingLocation];
    TRay ray = [camera pickRayX:location.x y:location.y];
    
    Picker* picker = [[windowController document] picker];
    PickingHitList* hitList = [picker pickObjects:&ray filter:filter];
    
    [activeDndTool concludeDragOperation:sender ray:&ray hits:hitList];
    activeDndTool = nil;
}

- (void)toggleTool:(Tool *)theTool index:(int)theIndex {
    Tool* modalReceiver = nil;
    if (modalReceiverIndex != -1) {
        modalReceiver = [receiverChain objectAtIndex:modalReceiverIndex];
        [modalReceiver deactivated:lastEvent ray:&lastRay hits:[self currentHits]];
        [receiverChain removeObjectAtIndex:modalReceiverIndex];
        modalReceiverIndex = -1;
    }

    if (theTool != modalReceiver) {
        modalReceiverIndex = theIndex;
        [theTool activated:lastEvent ray:&lastRay hits:[self currentHits]];
        [receiverChain insertObject:theTool atIndex:modalReceiverIndex];
    }
}

- (void)toggleDragVertexTool {
    [self toggleTool:dragVertexTool index:1];
}

- (void)toggleDragEdgeTool {
    [self toggleTool:dragEdgeTool index:1];
}

- (void)toggleDragFaceTool {
    [self toggleTool:dragFaceTool index:1];
}

- (void)toggleClipTool {
    [self toggleTool:clipTool index:1];
}

- (Tool *)currentModalTool {
    if (modalReceiverIndex == -1)
        return nil;
    return [receiverChain objectAtIndex:modalReceiverIndex];
}

- (ClipTool *)clipTool {
    return clipTool;
}

- (NSPoint)menuPosition {
    return menuPosition;
}

- (PickingHitList *)currentHits {
    if (currentHits == nil) {
        Picker* picker = [[windowController document] picker];
        currentHits = [[picker pickObjects:&lastRay filter:filter] retain];
        
        SelectionManager* selectionManager = [windowController selectionManager];
        if ([selectionManager mode] == SM_BRUSHES) {
            [picker pickCloseFaces:&lastRay brushes:[selectionManager selectedBrushes] maxDistance:10 hitList:currentHits];
            [picker pickVertices:&lastRay brushes:[selectionManager selectedBrushes] handleRadius:3 hitList:currentHits filter:filter];
        }
    } 
    
    return currentHits;
}

@end

File: Core/Source/Controller/InputController.cpp
/*
 Copyright (C) 2010-2012 Kristian Duske
 
 This file is part of TrenchBroom.
 
 TrenchBroom is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 TrenchBroom is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with TrenchBroom.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "InputController.h"
#include "Controller/Camera.h"
#include "Controller/CameraTool.h"
#include "Controller/CreateBrushTool.h"
#include "Controller/DragEntityTargetTool.h"
#include "Controller/DragTextureTargetTool.h"
#include "Controller/Editor.h"
#include "Controller/MoveEdgeTool.h"
#include "Controller/MoveFaceTool.h"
#include "Controller/MoveObjectTool.h"
#include "Controller/MoveVertexTool.h"
#include "Controller/SelectionTool.h"
#include "Controller/ResizeBrushTool.h"
#include "Model/Map/Map.h"
#include "Model/Map/Picker.h"
#include "Model/Preferences.h"
#include "Model/Selection.h"

namespace TrenchBroom {
    namespace Controller {
        void InputController::updateHits() {
            if (m_currentEvent.pickResults != NULL)
                delete m_currentEvent.pickResults;
            
            Model::Picker& picker = m_editor.map().picker();
            Camera& camera = m_editor.camera();
            m_currentEvent.ray = camera.pickRay(m_currentEvent.mouseX, m_currentEvent.mouseY);
            m_currentEvent.pickResults = picker.pick(m_currentEvent.ray, m_editor.filter());

            for (unsigned int i = 0; i < m_receiverChain.size(); i++)
                m_receiverChain[i]->updateHits(m_currentEvent);
        }

        void InputController::updateMousePos(float x, float y) {
            m_currentEvent.deltaX = x - m_currentEvent.mouseX;
            m_currentEvent.deltaY = y - m_currentEvent.mouseY;
            m_currentEvent.mouseX = x;
            m_currentEvent.mouseY = y;
        }

        void InputController::toggleModalTool(const ToolPtr& tool, unsigned int index) {
            ToolPtr modalTool;
            if (m_modalReceiverIndex != -1) {
                modalTool = m_receiverChain[m_modalReceiverIndex];
                modalTool->deactivated(m_currentEvent);
                m_receiverChain.erase(m_receiverChain.begin() + m_modalReceiverIndex);
                m_modalReceiverIndex = -1;
            }
            
            if (tool != modalTool) {
                m_modalReceiverIndex = index;
                m_receiverChain.insert(m_receiverChain.begin() + m_modalReceiverIndex, tool);
                tool->activated(m_currentEvent);
            }
        }
        
        bool InputController::modalToolActive(const ToolPtr& tool) {
            if (m_modalReceiverIndex == -1)
                return false;
            return m_receiverChain[m_modalReceiverIndex] == tool;
        }

        InputController::InputController(Editor& editor) : m_editor(editor), m_currentDragInfo(DragInfo(m_currentEvent)), m_modalReceiverIndex(-1) {
            ToolPtr cameraTool = ToolPtr(new CameraTool(m_editor));
            ToolPtr selectionTool = ToolPtr(new SelectionTool(m_editor));
            ToolPtr moveObjectTool = ToolPtr(new MoveObjectTool(m_editor));
            ToolPtr createBrushTool = ToolPtr(new CreateBrushTool(m_editor));
            ToolPtr resizeBrushTool = ToolPtr(new ResizeBrushTool(m_editor));
            
            m_receiverChain.push_back(cameraTool);
            m_receiverChain.push_back(resizeBrushTool);
            m_receiverChain.push_back(selectionTool);
            m_receiverChain.push_back(moveObjectTool);
            m_receiverChain.push_back(createBrushTool);
            
            m_dragButton = Tool::TB_MB_NONE;
            m_dragScrollReceiver = ToolPtr();

            m_moveVertexTool = ToolPtr(new MoveVertexTool(m_editor));
            m_moveEdgeTool = ToolPtr(new MoveEdgeTool(m_editor));
            m_moveFaceTool = ToolPtr(new MoveFaceTool(m_editor));
            
            DragTextureTargetTool* dragTextureTargetTool = new DragTextureTargetTool(m_editor);
            DragEntityTargetTool* dragEntityTargetTool = new DragEntityTargetTool(m_editor);
            
            m_dragTargetTools["Texture"] = DragTargetToolPtr(dragTextureTargetTool);
            m_dragTargetTools["Entity"] = DragTargetToolPtr(dragEntityTargetTool);
            
            m_currentEvent.mouseX = 0;
            m_currentEvent.mouseY = 0;
            m_currentEvent.deltaX = 0;
            m_currentEvent.deltaY = 0;
        }
        
        InputController::~InputController() {
            if (m_currentEvent.pickResults != NULL)
                delete m_currentEvent.pickResults;
        }
        
        void InputController::toggleMoveVertexTool() {
            toggleModalTool(m_moveVertexTool, 1);
        }

        void InputController::toggleMoveEdgeTool() {
            toggleModalTool(m_moveEdgeTool, 1);
        }
        
        void InputController::toggleMoveFaceTool() {
            toggleModalTool(m_moveFaceTool, 1);
        }
        
        bool InputController::moveVertexToolActive() {
            return modalToolActive(m_moveVertexTool);
        }
        
        bool InputController::moveEdgeToolActive() {
            return modalToolActive(m_moveEdgeTool);
        }
        
        bool InputController::moveFaceToolActive() {
            return modalToolActive(m_moveFaceTool);
        }
        
        
        bool InputController::key(wchar_t c) {
            return false;
        }

        void InputController::modifierKeyDown(Tool::EModifierKeys modifierKey) {
            m_currentEvent.modifierKeys |= modifierKey;
            updateHits();
            
            ToolList chain = m_receiverChain;
            for (unsigned int i = 0; i < chain.size(); i++)
                chain[i]->modifierKeyChanged(m_currentEvent);
        }
        
        void InputController::modifierKeyUp(Tool::EModifierKeys modifierKey) {
            m_currentEvent.modifierKeys &= ~modifierKey;
            updateHits();
            
            ToolList chain = m_receiverChain;
            for (unsigned int i = 0; i < chain.size(); i++)
                chain[i]->modifierKeyChanged(m_currentEvent);
        }
        
        bool InputController::mouseDown(Tool::EMouseButton mouseButton, float x, float y) {
            m_currentEvent.mouseButton = mouseButton;
            updateMousePos(x, y);
            updateHits();
            
            ToolList chain = m_receiverChain;
            for (unsigned int i = 0; i < chain.size(); i++) {
                if (chain[i]->mouseDown(m_currentEvent)) {
                    m_mouseDownReceiver = chain[i];
                    return true;
                }
            }
            
            return false;
        }
        
        bool InputController::mouseUp(Tool::EMouseButton mouseButton, float x, float y) {
            m_currentEvent.mouseButton = mouseButton;
            updateMousePos(x, y);
            updateHits();
            
            bool handled = false;
            if (m_currentEvent.mouseButton == m_dragButton) {
                if (m_dragScrollReceiver.get() != NULL)
                    m_dragScrollReceiver->endDrag(m_currentEvent);
                if (m_mouseDownReceiver.get() != NULL && m_mouseDownReceiver != m_dragScrollReceiver)
                    m_mouseDownReceiver->mouseUp(m_currentEvent);
                m_dragScrollReceiver = ToolPtr();
                m_dragButton = Tool::TB_MB_NONE;
                handled = true;
            } else {
                ToolList chain = m_receiverChain;
                for (unsigned int i = 0; i < chain.size() && !handled; i++)
                    if (chain[i]->mouseUp(m_currentEvent))
                        handled = true;
            }
            
            m_mouseDownReceiver = ToolPtr();
            m_currentEvent.mouseButton = Tool::TB_MB_NONE;
            return handled;
        }
        
        void InputController::mouseMoved(float x, float y) {
            if (m_currentEvent.mouseButton != Tool::TB_MB_NONE && m_dragButton == Tool::TB_MB_NONE) {
                m_dragButton = m_currentEvent.mouseButton;
                ToolList chain = m_receiverChain;
                for (unsigned int i = 0; i < chain.size(); i++) {
                    if (chain[i]->beginDrag(m_currentEvent)) {
                        m_dragScrollReceiver = chain[i];
                        break;
                    }
                }
            }

            updateMousePos(x, y);
            updateHits();
            
            if (m_dragButton != Tool::TB_MB_NONE && m_dragScrollReceiver.get() != NULL) {
                if (!m_dragScrollReceiver->drag(m_currentEvent)) {
                    m_dragScrollReceiver = ToolPtr();
                    m_mouseDownReceiver = ToolPtr();
                }
            }
            
            if (m_dragButton == Tool::TB_MB_NONE || m_dragScrollReceiver.get() == NULL) {
                ToolList chain = m_receiverChain;
                for (unsigned int i = 0; i < chain.size(); i++)
                    chain[i]->mouseMoved(m_currentEvent);
            }
        }
        
        void InputController::scrolled(float dx, float dy) {
            m_currentEvent.scrollX = dx;
            m_currentEvent.scrollY = dy;
            updateHits();
            
            if (m_dragScrollReceiver != NULL) {
                m_dragScrollReceiver->scrolled(m_currentEvent);
            } else {
                ToolList chain = m_receiverChain;
                for (unsigned int i = 0; i < chain.size(); i++)
                    if (chain[i]->scrolled(m_currentEvent))
                        break;
            }
        }

        bool InputController::dragEnter(const std::string& name, void* payload, float x, float y) {
            m_currentDragInfo.name = name;
            m_currentDragInfo.payload = payload;
            updateHits();

            bool overlayVisible = true;
            for (DragTargetToolMap::iterator it = m_dragTargetTools.begin(); it != m_dragTargetTools.end(); ++it) {
                it->second->deactivate(m_currentDragInfo);
                if (it->first == name)
                    overlayVisible = it->second->activate(m_currentDragInfo);
            }
        
            return overlayVisible;
        }
        
        void InputController::dragLeave(const std::string& name, void* payload) {
            m_currentDragInfo.name = name;
            m_currentDragInfo.payload = payload;
            updateHits();

            for (DragTargetToolMap::iterator it = m_dragTargetTools.begin(); it != m_dragTargetTools.end(); ++it)
                it->second->deactivate(m_currentDragInfo);
        }
        
        bool InputController::dragMove(const std::string& name, void* payload, float x, float y) {
            m_currentDragInfo.name = name;
            m_currentDragInfo.payload = payload;
            updateHits();

            DragTargetToolMap::iterator it = m_dragTargetTools.find(name);
            if (it == m_dragTargetTools.end())
                return true;
            
            return it->second->move(m_currentDragInfo);
        }

        bool InputController::acceptDrag(const std::string& name, void* payload) {
            m_currentDragInfo.name = name;
            m_currentDragInfo.payload = payload;
            updateHits();
            
            DragTargetToolMap::iterator it = m_dragTargetTools.find(name);
            if (it == m_dragTargetTools.end())
                return false;
            
            return it->second->accepts(m_currentDragInfo);
        }
        
        bool InputController::handleDrop(const std::string& name, void* payload, float x, float y) {
            m_currentDragInfo.name = name;
            m_currentDragInfo.payload = payload;
            updateHits();
            
            DragTargetToolMap::iterator it = m_dragTargetTools.find(name);
            if (it == m_dragTargetTools.end())
                return false;
            
            return it->second->drop(m_currentDragInfo);
        }
    }
}
